File: apps/server/openhive/auth/pkce.py
"""PKCE helpers (RFC 7636)."""

from __future__ import annotations

import base64
import hashlib
import secrets
from dataclasses import dataclass


@dataclass
class PKCEChallenge:
    code_verifier: str
    code_challenge: str
    state: str


def generate() -> PKCEChallenge:
    code_verifier = _urlsafe_b64(secrets.token_bytes(32))
    digest = hashlib.sha256(code_verifier.encode()).digest()
    code_challenge = _urlsafe_b64(digest)
    state = _urlsafe_b64(secrets.token_bytes(16))
    return PKCEChallenge(code_verifier, code_challenge, state)


def _urlsafe_b64(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).decode().rstrip("=")

File: apps/web/lib/server/auth/pkce.ts
/**
 * PKCE helpers (RFC 7636). Ports apps/server/openhive/auth/pkce.py.
 */

import crypto from 'node:crypto'

interface PKCEChallenge {
  code_verifier: string
  code_challenge: string
  state: string
}

function urlsafeB64(raw: Buffer): string {
  return raw.toString('base64').replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '')
}

export function generatePkce(): PKCEChallenge {
  const codeVerifier = urlsafeB64(crypto.randomBytes(32))
  const digest = crypto.createHash('sha256').update(codeVerifier).digest()
  const codeChallenge = urlsafeB64(digest)
  const state = urlsafeB64(crypto.randomBytes(16))
  return { code_verifier: codeVerifier, code_challenge: codeChallenge, state }
}
